lr_face/data_providers.py

```diff
@@ -236,4 +260,37 @@ def make_pairs(data: ImageWithIds) -> PairsWithIds:
 
         imgs_prev = imgs
         img_ids_prev = img_ids
-    return PairsWithIds(pairs=pairs, is_same_source=same_different_source, pair_ids=ids)
+    return PairsWithIds(pairs=pairs, is_same_source=same_different_source,
+                        pair_ids=ids)
+
+
+def make_triplets(data: ImageWithIds) -> List[Triplet]:
+    unique_identities = set(data.person_ids)
+    if len(unique_identities) < 2:
+        raise ValueError(
+            "Can't make triplets if there are fewer than 2 unique identities.")
+
+    # Pair the images with their identity.
+    images_with_identity: List[Tuple[np.ndarray, int]] = \
+        list(zip(data.images, data.person_ids))
+
+    # Convert data to a more pleasant format: a mapping from identities to a
+    # list of all images with that identity.
+    images_grouped_by_identity: Dict[int, List[np.ndarray]] = {
+        identity: [t[0] for t in group] for identity, group in groupby(
+            sorted(images_with_identity, key=lambda t: t[1]),
+            key=lambda t: t[1]
```

Comment:
Dit is wel echt magisch :D. Maar ik kan maar net volgen wat er gebeurt, ik weet niet of het mogelijk is om het iets verbozer/anders op te schrijven dat het wat beter leesbaar is?

Reply:
Ik heb nu veranderd naar:

```python
    images_grouped_by_identity = defaultdict(list)
    for image, identity in images_with_identity:
        images_grouped_by_identity[identity].append(image)
```